includes/class-sensei-learner.php

```diff
@@ -111,6 +114,63 @@ public function delete_all_user_activity( $user_id = 0 ) {
 		return $dataset_changes;
 	}
 
+	/**
+	 * Remove duplicate progress comments to mitigate duplicate enrollment issue.
+	 *
+	 * Hooked into sensei_log_activity_after
+	 *
+	 * @since 3.7.0
+	 * @access private
+	 *
+	 * @param array $args
+	 */
+	public function remove_duplicate_progress( $args ) {
+		add_action(
+			'shutdown',
+			function() use ( $args ) {
+				global $wpdb;
+
+				// Get progress comments, but the first one, which match the context conditions (they should not exist).
+				// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching, WordPress.DB.PreparedSQL.NotPrepared
+				$comment_ids = $wpdb->get_col(
+					$wpdb->prepare(
+						"SELECT comment_ID
+							FROM $wpdb->comments
+							WHERE comment_post_ID = %d
+								AND user_id = %d
+								AND comment_type = %s
+							ORDER BY comment_ID
+							LIMIT 1, 99",
+						$args['post_id'],
+						$args['user_id'],
+						$args['type']
+					),
+					0
+				);
+
+				if ( ! empty( $comment_ids ) ) {
+					$serialized_comment_ids = implode( ',', $comment_ids );
+
+					sensei_log_event(
+						'remove_duplicate_progress_comments',
+						[
+							'post_id'     => $args['post_id'],
+							'user_id'     => $args['user_id'],
+							'type'        => $args['type'],
+							'comment_ids' => $serialized_comment_ids,
+						]
+					);
+
+					$wpdb->query( "DELETE FROM $wpdb->comments WHERE comment_ID IN ( " . $serialized_comment_ids . ' )' );
+					$wpdb->query( "DELETE FROM $wpdb->commentmeta WHERE comment_id IN ( " . $serialized_comment_ids . ' )' );
```

Comment:
The `$serialized_comment_ids` string seems like it should be safe, but this code still opens up a theoretical (if unlikely) attack vector, since the comment IDs come from the database.

What about a strategy [like this one](https://stackoverflow.com/a/10634225)? Something like:

```php
$format_comment_ids = implode( ', ', array_fill( 0, count( $comment_ids ), '%s' ) );
$sql = "DELETE FROM $wpdb->comments WHERE comment_ID IN ( $format_comment_ids )";

$prepared = call_user_func_array( [ $wpdb, 'prepare' ], array_merge( [ $sql ], $comment_ids ) );
$wpdb->query( $prepared );
```

Reply:
Good catch! Changed here: https://github.com/Automattic/sensei/pull/3855/commits/90b9969b327ab36a9dfb979e05eda67a824a5479